extensions/amp-live-list/0.1/poller.js

```diff
@@ -0,0 +1,201 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {exponentialBackoff} from '../../../src/exponential-backoff';
+
+
+/**
+ * Polling state of a Poller instance.
+ * @enum {string}
+ */
+export const PollerState = {
+  /**
+   * The poller is running at a the set interval.
+   */
+  RUNNING: 'running',
+
+  /**
+   * The poller is running but is doing exponential backoff (using
+   * the original interval value as base) every tick.
+   */
+  BACKEDOFF: 'backedoff',
+
+  /**
+   * The poller is not running and not executing any work.
+   */
+  STOPPED: 'stopped',
+};
+
+
+/**
+ * Poller with backoff functionality.
+ */
+export class Poller {
+
+  constructor(win, wait, work) {
+    /** @private @const {!Window} */
+    this.win = win;
+
+    /** @private {number} */
+    this.wait_ = wait;
+
+    /** @private {function(): !Promise} */
+    this.work_ = work;
+
+    /** @private {?number} */
+    this.lastTimeoutId_ = null;
+
+    /** @private {?number} */
+    this.lastBackoffTimeoutId_ = null;
+
+    /** @private {?function(): number} */
+    this.backoffPoller_ = null;
+
+    /** @private {!PollerState} */
+    this.state_ = PollerState.STOPPED;
+
+    /**
+     * Mostly for testing purposes.
+     * @private {!Promise}
+     */
+    this.lastWorkPromise_ = null;
+  }
+
+  /**
+   * @return {function(): number}
+   * @private
+   */
+  getBackoffPoller_() {
+    return exponentialBackoff();
+  }
+
+  /**
+   * @return {!PollerState}
+   */
+  getState() {
+    return this.state_;
+  }
+
+  /**
+   * Initalize any work needed to start polling.
+   */
+  start() {
+    if (this.lastTimeoutId_) {
+      return;
+    }
+
+    if (this.lastBackoffTimeoutId_) {
+      this.win.clearTimeout(this.lastBackoffTimeoutId_);
+      this.lastBackoffTimeoutId_ = null;
+    }
+
+    // We reset the backoff poller
+    this.backoffPoller_ = null;
+    this.state_ = PollerState.RUNNING;
+    this.poll_();
+  }
+
+  /**
+   * Clear timers and set correct stop state.
+   */
+  stop() {
+    if (this.state_ == PollerState.STOPPED) {
+      return;
+    }
+
+    this.state_ = PollerState.STOPPED;
+    if (this.lastTimeoutId_) {
+      this.win.clearTimeout(this.lastTimeoutId_);
+      this.lastTimeoutId_ = null;
+    }
+    if (this.lastBackoffTimeoutId_) {
+      this.win.clearTimeout(this.lastBackoffTimeoutId_);
+      this.lastBackoffTimeoutId_ = null;
+      this.backoffPoller_ = null;
+    }
+  }
+
+  /**
+   * Assign correct state and call internal recursive backoffPoll_.
+   * @private
+   */
+  backoff_() {
+    if (this.lastBackoffTimeoutId_) {
+      return;
+    }
+    if (!this.backoffPoller_) {
+      this.backoffPoller_ = this.getBackoffPoller_();
+    }
+    this.stop();
```

Comment:
Does this make sense to call `stop` here? The `lastBackoffTimeoutId_` you already check above and we know that the current normal timer (`lastTimeoutId_`) is not running since this method is called when it fails. As it stands, however, the API allows to call `start()` method while in back off and it will run a normal poll. I don't think you intended that. This could probably be simply fixed by switching `start` to do `if (this.state_ != STOPED) {return}` from current `if (lastTimeoutId_) {return}`